framework/include/utils/MooseUtils.h

```diff
@@ -983,6 +984,48 @@ buildRequiredMaterials(const Consumers & mat_consumers,
                        const std::vector<std::shared_ptr<MaterialBase>> & mats,
                        const bool allow_stateful);
 
+/**
+ * Utility class template for a semidynamic vector with a maximum size N
+ * and a chosen dynamic size. This container avoids heap allocation and
+ * is meant as a replacement for small local std::vector variables.
+ */
+template <typename T, std::size_t N>
+class SemidynamicVector
```

Comment:
This looks almost exactly like `MetaPhysicL::DynamicStdArrayWrapper` with some slight improvements in a couple of places, e.g. I prefer your implementation of `end()` because it removes an operation. Perhaps we could improve and reuse the current class? I know that this will make it slower to get it available in MOOSE...

Reply:
Yeah, good idea. Swapping the use of this class out would involve changing 3 typedefs. I can check if `MetaPhysicL::DynamicStdArrayWrapper` works out of the box. If so I'll scrap my implementation and a PR to change end() could be made at any point in time.

Reply:
You could alias `DynamicStdArrayWrapper` to a name you like better

Reply:
is there maybe some advanced reason for using `_data.end() - (N - _dynamic_n);` rather than `_data.begin() + _dynamic_n;` ?

Reply:
I do not think so, but I suppose that I cannot say for sure until we try it out. It's possible that past Alex did this for a good reason, but current Alex thinks he probably just overlooked this.

Reply:
Did it not work out of the box? I feel as if you're getting pretty far down the road here...

Reply:
Sorry, I haven't even touched that part again yet, there was so much other stuff on this PR that kept me busy :-). Any comments regarding my last comment above?

Reply:
By the way, is the `NType` a clever way of avoiding non-type template parameters?

Reply:
`DynamicStdArrayWrapper::DynamicStdArrayWrapper(std::size_t n)` does not exist. And The MooseADWrapper template that works for `std::array` doesn't apply (because of the `NType`  vs. `std::size_t` template argument). All that could be worked around. But color me concerned about the `MooseADWrapper<DynamicStdArrayWrapper>::type` issue.

Reply:
I suppose I could change my class to derive from DynamicStdArrayWrapper and implement that ctor like

```
template <std::size_t N>
class SemidynamicVector : public DynamicStdArrayWrapper<NWrapper<N>>
{
  SemidynamicVector(std::size_t n) : _dynamic_n(n) {}
  // possibly some 'using' stuff
};
```

Reply:
Yeah, good point. `std::vector` uses _zero initialization_, while _std::array_ uses _default initialization_ (meaning no initialization for built-ins). If the class is supposed to be a drop in for _std::vector_, we should add an `assign` call to the ctor.

Reply:
`ADReal` is a `DualNumber<Real, SemiDynamicSparseNumberArray<Real, dof_id_type, NWrapper<n>>, true>`, in other words an instantiation of a `template <typename, typename, bool>` so there are no dangers of your template specialization matching `DualNumber` itself

Reply:
sure, but I could still at least derive for your container. It is a bit more comprehensive than my implementation anyways with reverse iterators and move construction.

Reply:
At this point I would say up to you. I'll accept either choice

Reply:
Ok, I have a derived version ready, but `_data` is private, so I cannot add my initialization. I'll just leave it at _default initialization_ for now and put in a PR to make the members protected, if that's fine with you.

Reply:
sounds good to me!